Check existence revalidates on bad status

jmartin-sul opened the pull request:
~~pretty close, added new tests, still need to:~~
- [x] move most of the new tests to a shared example, except for the ones that deal with status becoming `EXPECTED_VERS_NOT_FOUND_ON_STORAGE_STATUS`, since the conditions for that differ between `check_existence` and `confirm_version`.
- [x] re-implement the `EXPECTED_VERS_NOT_FOUND_ON_STORAGE_STATUS` tests appropriately for `confirm_version` (e.g. add a test where the version increases, which is fine by `check_existence`, but not `confirm_version`).
- [x] call the shared example on `check_existence` and `confirm_version` in the appropriate place in the respective spec files.

paired with @tingulfsen for most of this.

supersedes #438 

closes #431

closes #433

jmartin-sul commented:
i think this is ready for review.  in the process of trying to figure out why the tests i initially added as skipped in the second commit weren't working, i saw what seemed like a couple convenient/easy refactorings that'd help me out.

the first was to have the method that was validating the object in the case of an already bad status also do the validation and status setting if the incoming version is unexpected compared to the catalog.  with a bit of extra mocking (since it triggered slightly more frequent object validation), this allowed enabling the previously skipped tests.  it also allowed a bit of simplification in the status setting that `check_existence` does.

the second was to centralize the setting of the `ran_moab_validation` flag.  i think there was a moment where that felt directly related to the rest of this PR's work.  i'm not sure it actually is, but it did feel like an improvement in maintainability in the vein of some other refactoring we've done (e.g. turning result codes into a centrally managed list).

i tried to build this up with a logical ordering of commits, with the ones i thought might be more controversial saved for later.  but fixing the skipped tests seemed important, because they seemed to be testing reasonable expectations.  and hopefully i wasn't way off on the validation flag one.

probably easier to review commit by commit.

ndushay commented on spec/services/preserved_object_handler_check_exist_spec.rb:
```diff
@@ -386,6 +376,38 @@
+            po_handler.check_existence
+            expect(pc.reload.status).to eq PreservedCopy::OK_STATUS
+          end
+          it 'had EXPECTED_VERS_NOT_FOUND_ON_STORAGE_STATUS, seems to have an acceptable version now' do
```
"seems to have" ?   hahaha - it's a test - I think we could be assertive here.  Mostly added this comment to amuse myself, not because I think changing the text is important.

jmartin-sul replied:
"gee, sure looks like it has an acceptable version now, but really who are we to judge what's an acceptable version?"

ndushay commented on spec/services/preserved_object_handler_confirm_version_spec.rb:
```diff
@@ -124,6 +124,31 @@
         end
       end
 
```
question:  for "confirm_version" do we now have the capacity to easily report when incoming version > catalog version + 1?    That is, if incoming is 4 and catalog is 2 ... does the code now support us detecting this case?  Asking because it pertains to ticket #451

jmartin-sul replied:
i don't think this work makes it easier (or harder).  we'd still have to differentiate the case where the version is greater, since we only look for whether the versions are equal or not in `confirm_version`.

and while i don't think this PR makes that differentiation easier in `check_existence` either, that differentiation is already easier in `check_existence` than in `confirm_version`, since `check_existence` already differentiates the `incomnig_version > pres_copy.version` case from other version mismatches (in the `incoming_version > pres_copy.version` condition in `check_existence`, we could just see whether the difference is greater than one, and return different result codes accordingly).

ndushay commented on spec/services/shared_examples_preserved_object_handler.rb:
```diff
@@ -474,3 +474,106 @@
+    po_handler.send(method_sym)
+    expect(pc.reload.status).to eq PreservedCopy::INVALID_MOAB_STATUS
+  end
+  it 'had EXPECTED_VERS_NOT_FOUND_ON_STORAGE_STATUS, seems to have an acceptable version now' do
```
another "seems to"

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -76,7 +76,7 @@ def check_existence
           incoming_vers_gt_pc_vers = (incoming_version > pres_copy.version)
           found_expected_version = (incoming_vers_eq_pc_vers || incoming_vers_gt_pc_vers)
-          check_status_if_not_ok(pres_copy, found_expected_version)
+          check_status_if_not_ok_or_unexpected_version(pres_copy, found_expected_version)
```
I'm wondering if the code is clearer if you take each comparator relationship separately below and call `check_status_if_not_ok_or_unexpected_version(pres_copy, false)` or true as the case may be.  

The reason I suggest this is the comment with line 100 about ran_moab_validation setting -- it feels like there is some slight of hand going on now ...

But have yet to look at the whole -- still going commit by commit ...

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -272,14 +267,19 @@ def update_pc_invalid_moab
 
-  def check_status_if_not_ok(pres_copy, found_expected_version)
-    if pres_copy.status != PreservedCopy::OK_STATUS
+  # TODO: what if this checked status in the case of version mismatcch also?  would that simplify
```
mismatccccch

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -272,14 +267,19 @@ def update_pc_invalid_moab
-    if pres_copy.status != PreservedCopy::OK_STATUS
+  # TODO: what if this checked status in the case of version mismatcch also?  would that simplify
+  # check_existence and/or confirm_version in their handling after calling this?  could they just
+  # deal with the result code wranging after calling this if it also checked status on vers mismatch?
```
The question you raise here is worthy - I wonder if it should be a refactor ticket rather than a comment here?  Or in addition?  Or maybe worth no more than 1 hour's experimentation?

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -272,14 +267,19 @@ def update_pc_invalid_moab
 
   # given a PreservedCopy instance and whether the caller found the expected version of it on disk, this will perform
   # other validations of what's on disk, and will update the status accordingly
+  # FIXME: this method needs to update validation timestamps, and tests need to get fixed to reflect that
```
I hope this is addressed in a later commit ...

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -272,14 +267,19 @@ def update_pc_invalid_moab
+  # TODO: what if this checked status in the case of version mismatcch also?  would that simplify
+  # check_existence and/or confirm_version in their handling after calling this?  could they just
+  # deal with the result code wranging after calling this if it also checked status on vers mismatch?
+  def check_status_if_not_ok_or_unexpected_version(pres_copy, found_expected_version)
```
such a long name.  How about `validate_for_status`?  or `validate_status` or `status_validation` or ...

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -174,19 +176,25 @@ def update_version
-        error_hash.each_value { |msg| moab_error_msgs << msg }
+    # validation is a bit expensive, so memoize the result.  keeps us from having to worry about a
+    # performance hit if a caller wants to check whether there are validation errors from more than
+    # one conditional statement.
```
i think you can leave out this comment - memoization is a common practice and devs know why it is used.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -73,33 +73,32 @@ def check_existence
             handler_results.add_result(PreservedObjectHandlerResults::ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_object.class.name)
-            ran_moab_validation = true
+            # we're not too worried if the version increased, but we normally expect it to be kept current
+            # by calls from ingest robots, so validate to make sure all's well.
```
not sure this comment is giving us anything -- i would prefer to read the code to figure it out.  ("so if the incoming is greater then it validates ...")

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -73,33 +73,32 @@ def check_existence
+          incoming_vers_eq_pc_vers = (incoming_version == pres_copy.version)
+          incoming_vers_gt_pc_vers = (incoming_version > pres_copy.version)
+          found_expected_version = (incoming_vers_eq_pc_vers || incoming_vers_gt_pc_vers)
+          check_status_if_not_ok_or_unexpected_version(pres_copy, found_expected_version)
```
what if you put `check_status_if_not_ok_or_unexpected_version(pres_copy, true)` (or false)  in each of the if clauses below?   I have a preference for actually seeing "incoming_version > pres_copy.version" over the variable.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -73,33 +73,32 @@ def check_existence
+              update_preserved_copy_version_etc(pres_copy, incoming_version, incoming_size)
               pres_object.current_version = incoming_version
               update_db_object(pres_object)
+              # don't have to set status here: check_status_if_not_ok_or_unexpected_version should've corrected a bad status
```
putting the call to `check_status_if_not_ok_or_unexpected_version(pres_copy, true)` within the if-elsif-else statement makes this comment unnecessary.   As it stands, I feel like it's calling out some slight of hand.

Or maybe simply removing this comment works.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -261,19 +271,51 @@ def update_pc_invalid_moab
 
-  def update_pc_unexpected_version(pres_copy, pres_object, new_status, ran_moab_validation)
+  def check_status_if_not_ok_or_unexpected_version(pres_copy, found_expected_version)
+    # if pres_copy.status != PreservedCopy::OK_STATUS
```
I think this comment can go.  I think the unless reads clearly.

jmartin-sul replied:
yeah, that was a commented out line from a prior iteration, not an intentional comment, will remove

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -261,19 +271,51 @@ def update_pc_invalid_moab
+
+  # given a PreservedCopy instance and whether the caller found the expected version of it on disk, this will perform
+  # other validations of what's on disk, and will update the status accordingly
+  # FIXME: this method probably needs to update validation timestamps?  and tests would need to get fixed to reflect that.
```
i think this FIXME comment can now go away?

jmartin-sul replied:
was about to say i think it's still needed, but now i agree it can go.  i think i left that comment before doing the `ran_moab_validation` refactoring.  now that that flag gets set on the POH instance whenever validation is run, i agree that this should be able to go away, because the top-level callers to `set_status_as_seen_on_disk` both also call `update_pc_audit_timestamps` (which now reads the centrally set `ran_moab_validation` flag).

sorry if you'd already worked all that out, i wrote that down as much for my own clarification as anything.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -261,19 +271,51 @@ def update_pc_invalid_moab
+  # FIXME: this method probably needs to update validation timestamps?  and tests would need to get fixed to reflect that.
+  def set_status_as_seen_on_disk(pres_copy, found_expected_version)
+    # TODO: do the check that'd set PreservedCopy::ONLINE_MOAB_NOT_FOUND_STATUS
+    # TODO: nothing in the code seems to set ONLINE_MOAB_NOT_FOUND_STATUS atm, but everything we're checking right now is M2C and thus found by definition
```
I think the above two comments could be combined.  Maybe "TODO: when we implement C2M, this is likely where we'll set status to ..."

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -261,19 +271,51 @@ def update_pc_invalid_moab
+
+    # TODO: do the check that'd set FIXITY_CHECK_FAILED_STATUS
+
+    # should automatically get things out of VALIDITY_UNKNOWN_STATUS, because if we fall through to here, we can set status to ok
```
this comment could be removed as well.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -286,14 +328,16 @@ def confirm_online_version
       raise_rollback_if_pc_po_version_mismatch(pres_copy.version, pres_object.current_version)
 
-      if incoming_version == pres_copy.version
+      found_expected_version = (incoming_version == pres_copy.version)
```
again, I have a slight preference for `incoming_version == pres_copy.version` in the if statement.  Here, tho, the logic is simpler so my cognitive overload is lower.

SaravShah commented on app/services/preserved_object_handler.rb:
```diff
@@ -73,11 +73,16 @@ def check_existence
 
           raise_rollback_if_pc_po_version_mismatch(pres_copy.version, pres_object.current_version)
 
+          incoming_vers_eq_pc_vers = (incoming_version == pres_copy.version)
```
i do have a slight preference for the conditions in the if statement, but i dont mind it here.

SaravShah commented on app/services/preserved_object_handler.rb:
```diff
@@ -267,6 +272,36 @@ def update_pc_invalid_moab
     handler_results.remove_db_updated_results unless transaction_ok
   end
 
+  def check_status_if_not_ok(pres_copy, found_expected_version)
```
👍

SaravShah commented on app/services/preserved_object_handler.rb:
```diff
@@ -191,7 +189,11 @@ def moab_validation_errors
 
-  def create_db_objects(status, ran_moab_validation=false)
+  def ran_moab_validation?
+    @ran_moab_validation
```
👍

jmartin-sul replied:
ah, good question, hadn't thought about that explicitly... but i think what's happening is that if an instance var hasn't been set yet, and it gets read, `nil` will get returned, which is "falsey", so essentially yeah, it returns `false` for our use case.

but that's a little unclear, so i'll change the method to do `@ran_moab_validation ||= false`.  which'll explicitly init `@ran_moab_validation` to false if it's still unset, or leave it alone if it's already there.  hopefully things read a bit clearer with that change.

ndushay replied:
good catch, @SaravShah and @jmartin-sul !

jmartin-sul commented:
i believe i've addressed all of the feedback from @ndushay and @SaravShah, should be ready for re-evaluation.

i just pushed follow on commits for the changes i made.  normally i'd be more of a stickler about a tidy commit history, but this isn't a terrible amount of churn, the squashing would be a bit of a pain, and this is probably a little easier for follow-on review anyway.  happy to cleanup if people would prefer that.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -261,19 +263,42 @@ def update_pc_invalid_moab
+  # other validations of what's on disk, and will update the status accordingly
+  def set_status_as_seen_on_disk(pres_copy, found_expected_version)
+    # TODO: when we implement C2M, we should add a check here to look for the object on disk, and set PreservedCopy::ONLINE_MOAB_NOT_FOUND_STATUS
+    # if needed.  or maybe the caller will have done that already, since it needs to try reading the moab to obtain incoming_version?
```
this assumes C2M will be in pohandler.  That is ... unclear at this time.   Should this comment stay?

jmartin-sul replied:
good point, i'd been assuming C2M would leverage POH methods that called this new method, but that seems less clear after the discussions this afternoon.  happy to remove this comment or make it appropriately vague.

ndushay replied:
I dunno what to do;  confusion reigns in what some what call my "mind"
